[SPARK-19825][R][ML] spark.ml R API for FPGrowth

zero323 opened the pull request:
## What changes were proposed in this pull request?

Adds SparkR API for FPGrowth: [SPARK-19825](https://issues.apache.org/jira/browse/SPARK-19825):

- `spark.fpGrowth` -model training.
- `freqItemsets` and `associationRules` methods with new corresponding generics.
- Scala helper: `org.apache.spark.ml.r. FPGrowthWrapper`
- unit tests.

## How was this patch tested?

Feature specific unit tests.

zero323 committed:
Add R API for FPGrowth

SparkQA commented:
**[Test build #73956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73956/testReport)** for PR 17170 at commit [`641fe70`](https://github.com/apache/spark/commit/641fe70362ad7460e85795a5a5aa58c2a990ebcf).

SparkQA commented:
**[Test build #73956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73956/testReport)** for PR 17170 at commit [`641fe70`](https://github.com/apache/spark/commit/641fe70362ad7460e85795a5a5aa58c2a990ebcf).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `    class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73956/
Test FAILed.

SparkQA commented:
**[Test build #73957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73957/testReport)** for PR 17170 at commit [`b53963a`](https://github.com/apache/spark/commit/b53963a898ee08f309443023c2ddf5ade6fdd2c5).

SparkQA commented:
**[Test build #73957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73957/testReport)** for PR 17170 at commit [`b53963a`](https://github.com/apache/spark/commit/b53963a898ee08f309443023c2ddf5ade6fdd2c5).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73957/
Test FAILed.

SparkQA commented:
**[Test build #73959 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73959/testReport)** for PR 17170 at commit [`021cd9b`](https://github.com/apache/spark/commit/021cd9b4dcaa19d4bcf912ab079c263bedc7889b).

SparkQA commented:
**[Test build #73959 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73959/testReport)** for PR 17170 at commit [`021cd9b`](https://github.com/apache/spark/commit/021cd9b4dcaa19d4bcf912ab079c263bedc7889b).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73959/
Test FAILed.

SparkQA commented:
**[Test build #73961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73961/testReport)** for PR 17170 at commit [`b198dfa`](https://github.com/apache/spark/commit/b198dfae88a3062b5966fd672a8f14f600e6fb32).

SparkQA commented:
**[Test build #73961 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73961/testReport)** for PR 17170 at commit [`b198dfa`](https://github.com/apache/spark/commit/b198dfae88a3062b5966fd672a8f14f600e6fb32).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73961/
Test FAILed.

SparkQA commented:
**[Test build #73962 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73962/testReport)** for PR 17170 at commit [`03789d6`](https://github.com/apache/spark/commit/03789d6890a65cddea6612fadc1bd75506939d5d).

SparkQA commented:
**[Test build #73962 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73962/testReport)** for PR 17170 at commit [`03789d6`](https://github.com/apache/spark/commit/03789d6890a65cddea6612fadc1bd75506939d5d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73962/
Test FAILed.

SparkQA commented:
**[Test build #73963 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73963/testReport)** for PR 17170 at commit [`d50f917`](https://github.com/apache/spark/commit/d50f917c7c749057b58b205f7694ba8caa1332ef).

SparkQA commented:
**[Test build #73963 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73963/testReport)** for PR 17170 at commit [`d50f917`](https://github.com/apache/spark/commit/d50f917c7c749057b58b205f7694ba8caa1332ef).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73963/
Test FAILed.

SparkQA commented:
**[Test build #73966 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73966/testReport)** for PR 17170 at commit [`6554384`](https://github.com/apache/spark/commit/65543840b347680123f5a478c0d8454b2a08482f).

SparkQA commented:
**[Test build #73966 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73966/testReport)** for PR 17170 at commit [`6554384`](https://github.com/apache/spark/commit/65543840b347680123f5a478c0d8454b2a08482f).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `    class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73966/
Test FAILed.

SparkQA commented:
**[Test build #74021 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74021/testReport)** for PR 17170 at commit [`0b34d03`](https://github.com/apache/spark/commit/0b34d03d658f1aba1807585041870a5e8c4264d9).

SparkQA commented:
**[Test build #74021 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74021/testReport)** for PR 17170 at commit [`0b34d03`](https://github.com/apache/spark/commit/0b34d03d658f1aba1807585041870a5e8c4264d9).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `    class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74021/
Test FAILed.

SparkQA commented:
**[Test build #74022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74022/testReport)** for PR 17170 at commit [`86f96a5`](https://github.com/apache/spark/commit/86f96a53a00d0b4013148d3060d0c3151b288d2f).

SparkQA commented:
**[Test build #74022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74022/testReport)** for PR 17170 at commit [`86f96a5`](https://github.com/apache/spark/commit/86f96a53a00d0b4013148d3060d0c3151b288d2f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `    class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74022/
Test PASSed.

SparkQA commented:
**[Test build #74030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74030/testReport)** for PR 17170 at commit [`6c0aea9`](https://github.com/apache/spark/commit/6c0aea9ffc66bf525f61efca14c8630dbb940d52).

SparkQA commented:
**[Test build #74030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74030/testReport)** for PR 17170 at commit [`6c0aea9`](https://github.com/apache/spark/commit/6c0aea9ffc66bf525f61efca14c8630dbb940d52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74030/
Test PASSed.

felixcheung commented on R/pkg/R/generics.R:
```diff
@@ -1420,6 +1420,17 @@ setGeneric("spark.posterior", function(object, newData) { standardGeneric("spark
+
+#' @rdname spark.fpGrowth
+#' @export
+setGeneric("freqItemsets", function(object) { standardGeneric("freqItemsets") })
```
we seems to follow the pattern `spark.something` - see LDA. do you think it makes sense here too?

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+#' }
+#' @note spark.fpGrowth since 2.2.0
+setMethod("spark.fpGrowth", signature(data = "SparkDataFrame"),
+          function(data, minSupport = 0.3, minConfidence = 0.8,
```
should it have `numPartitions`?

zero323 replied:
Done.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+#' @note spark.fpGrowth since 2.2.0
+setMethod("spark.fpGrowth", signature(data = "SparkDataFrame"),
+          function(data, minSupport = 0.3, minConfidence = 0.8,
+                   featuresCol = "features", predictionCol = "prediction") {
```
instead of `features` it should take a formula?

felixcheung replied:
we generally avoid allow setting `predictionCol` too

felixcheung replied:
about here-  thought?

zero323 replied:
To be honest I am not sure. If you think that setting `predictionCol` should  be disabled I am fine with that but I don't see how formulas could be useful here. `FPGrowth` doesn't really conform to the conventions used in other ML algorithms. It doesn't use vectors and fixed size buckets are unlikely to happen.

felixcheung replied:
I believe predictionCol param only allow you to change the name of the column - prediction is always still going to be there, no?

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+            }
+
+            jobj <- callJStatic("org.apache.spark.ml.r.FPGrowthWrapper", "fit",
+                                data@sdf, minSupport, minConfidence,
```
you may want to `as.numeric` on `minSupport`, `minConfidence` in case someone is passing in an integer and `callJStatic` would fail to match the wrapper method

zero323 replied:
Done.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+          function(object) {
+            jobj <- object@jobj
+            freqItemsets <- callJMethod(jobj, "freqItemsets")
+            dataFrame(freqItemsets)
```
It might make sense to do this in a single line:
```
dataFrame(callJMethod(object@jobj, "freqItemsets")
```

might be more readable that way. ditto with Association Rules below

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+#' @note FPGrowthModel since 2.2.0
+setClass("FPGrowthModel", slots = list(jobj = "jobj"))
+
+#' FPGrowth Model
```
could you use the long form name (eg. look at LDA) and drop the word "Model" which we avoid using

zero323 replied:
Do you mean `spark.FPGrowth`? I can but as far as I can tell all classes use `Model` suffix ([`GeneralizedLinearRegressionModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_regression.R#L33), [`GaussianMixtureModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_clustering.R#L32) [`LDAModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_clustering.R#L46) and so on) and none is using `spark` prefix.

Or do you mean `representation` instead of `slots`? I believe that representation is no longer recommended.

felixcheung replied:
I mean this 
https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L467
https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L316
which may or may not include the word model

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+#' @examples
+#' \dontrun{
+#' itemsets <- data.frame(features = c("a,b", "a,b,c", "c,d"))
+#' data <- selectExpr(createDataFrame(itemsets), "split(features, ',') as features")
```
instead of duplicating `createDataFrame`, set `itemsets <- createDataFrame(data.frame(features = c("a,b", "a,b,c", "c,d")))`

btw, do we have real data to use instead?

zero323 replied:
Yes, we do. Adjusted.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,144 @@
+#' 
+#' # Show frequent itemsets
+#' frequent_itemsets <- freqItemsets(model)
+#' showDF(frequent_itemsets)
```
collapse this to `head(freqItemsets(model))`

felixcheung commented on R/pkg/inst/tests/testthat/test_mllib_fpm.R:
```diff
@@ -0,0 +1,74 @@
+    freq = c(2, 2, 3, 3, 4)
+  )
+
+  expect_equivalent(expected_itemsets, collect(freqItemsets(model)))
```
don't repeat `freqItemsets(model)` - use `itemsets` from above

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala:
```diff
@@ -0,0 +1,84 @@
+import org.apache.hadoop.fs.Path
+import org.json4s._
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
```
do we need these?

zero323 replied:
We can skip `import org.json4s._` if won't do any parsing, but import org.json4s.jackson.JsonMethods._` provide both `render` and `compact` which are used to create JSON metadata.

felixcheung reviewed (changes requested):
thanks, I think a couple of big changes are in order. Let me know when it is ready for review again

SparkQA commented:
**[Test build #74111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74111/testReport)** for PR 17170 at commit [`1014902`](https://github.com/apache/spark/commit/10149026cd9ed2e33e7cf6a769bc48846e3519b1).

SparkQA commented:
**[Test build #74111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74111/testReport)** for PR 17170 at commit [`1014902`](https://github.com/apache/spark/commit/10149026cd9ed2e33e7cf6a769bc48846e3519b1).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74111/
Test FAILed.

SparkQA commented:
**[Test build #74113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74113/testReport)** for PR 17170 at commit [`eb39222`](https://github.com/apache/spark/commit/eb39222862eff2cffb6bcd6650c325ef9f82de4f).

SparkQA commented:
**[Test build #74114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74114/testReport)** for PR 17170 at commit [`bf26f79`](https://github.com/apache/spark/commit/bf26f793c478cc99c69f78ff28563772a0550699).

SparkQA commented:
**[Test build #74115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74115/testReport)** for PR 17170 at commit [`956a36a`](https://github.com/apache/spark/commit/956a36a47249e64825e03e4691d2b70646c84000).

SparkQA commented:
**[Test build #74113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74113/testReport)** for PR 17170 at commit [`eb39222`](https://github.com/apache/spark/commit/eb39222862eff2cffb6bcd6650c325ef9f82de4f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74113/
Test PASSed.

SparkQA commented:
**[Test build #74114 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74114/testReport)** for PR 17170 at commit [`bf26f79`](https://github.com/apache/spark/commit/bf26f793c478cc99c69f78ff28563772a0550699).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74114/
Test PASSed.

SparkQA commented:
**[Test build #74115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74115/testReport)** for PR 17170 at commit [`956a36a`](https://github.com/apache/spark/commit/956a36a47249e64825e03e4691d2b70646c84000).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74115/
Test PASSed.

felixcheung reviewed (commented):
https://github.com/apache/spark/pull/17170/files/6c0aea9ffc66bf525f61efca14c8630dbb940d52#r104594539

SparkQA commented:
**[Test build #74119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74119/testReport)** for PR 17170 at commit [`6be7f13`](https://github.com/apache/spark/commit/6be7f1322815c62a7f1259b8789b6e40f446f6c6).

SparkQA commented:
**[Test build #74123 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74123/testReport)** for PR 17170 at commit [`1949da3`](https://github.com/apache/spark/commit/1949da3082e9a720f1421bc2b977557ea199b4ef).

SparkQA commented:
**[Test build #74119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74119/testReport)** for PR 17170 at commit [`6be7f13`](https://github.com/apache/spark/commit/6be7f1322815c62a7f1259b8789b6e40f446f6c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74119/
Test PASSed.

SparkQA commented:
**[Test build #74123 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74123/testReport)** for PR 17170 at commit [`1949da3`](https://github.com/apache/spark/commit/1949da3082e9a720f1421bc2b977557ea199b4ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74123/
Test PASSed.

SparkQA commented:
**[Test build #74132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74132/testReport)** for PR 17170 at commit [`71f23ee`](https://github.com/apache/spark/commit/71f23eeb957f75a827b0a8498ca3f8d22ed76501).

SparkQA commented:
**[Test build #74132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74132/testReport)** for PR 17170 at commit [`71f23ee`](https://github.com/apache/spark/commit/71f23eeb957f75a827b0a8498ca3f8d22ed76501).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74132/
Test PASSed.

SparkQA commented:
**[Test build #74145 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74145/testReport)** for PR 17170 at commit [`3db1413`](https://github.com/apache/spark/commit/3db14134f2e30c47d801c9defd4b1081eb9010e6).

SparkQA commented:
**[Test build #74145 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74145/testReport)** for PR 17170 at commit [`3db1413`](https://github.com/apache/spark/commit/3db14134f2e30c47d801c9defd4b1081eb9010e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74145/
Test PASSed.

zero323 commented:
@felixcheung I think i addressed all the issues excluding `inputCol` and `predictionCol`. In general:

- Using `formula`as an input doesn't make sense in my opinion. 
- Personally I would allow users to set column names. Both `features` and `prediction` are a bit vague in the context of the algorithm.

felixcheung commented:
Sure - it's a single columns called `features` so I'm fine with it as a parameter.
I'm not sure about `inputCol` though, it's a different nomenclature in R that is different from the rest of mllib `features`.

zero323 commented:
I think that [ALS sets a precedence for using `somethingCol`](https://github.com/apache/spark/blob/master/R/pkg/R/mllib_recommendation.R#L86) but I don't like 'features" part here. Maybe `basketsCol`, what you think?

felixcheung commented:
To clarify with your example, with ALS we have userCol, ratingCol - these matches the API names in spark.ml, and I think we need to do the same here.

What don't you like about [`featuresCol` which is also in the Scala API](https://github.com/apache/spark/blob/0fe8020f3aaf61c9992b6bcc5dba7ae8f751bab7/mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala#L201)? `featuresCol` is really a standard trait with `HasFeaturesCol` used in different ml model across Spark.

zero323 commented:
My point is that whenever we use `HasFeaturesCol` it is expected that `featuresCol` is `ml.linalg.VectorUDT`. This creates nice and consistent API. 

If we use it for `FPGrowth` it is no longer so clear. Once / if `PrefixSpan` is implemented we'll get just another input format using the same `trait` / mixin. If you look at SO you'll see that some users  are already confused what is the expected input for ML algorithms and I believe this can make it even worse. Just saying...

felixcheung commented:
I can see your point, but renaming it only on the R side is not really addressing the issue.
Please feel free to open a JIRA on spark.ml FPGrowth and start a discussion there.

zero323 commented:
You're right of course: [SPARK-19899](https://issues.apache.org/jira/browse/SPARK-19899).

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+#' 
+#' A parallel FP-growth algorithm to mine frequent itemsets. The algorithm is described in
+#' Li et al., PFP: Parallel FP-Growth for Query
+#' Recommendation <\url{http://dx.doi.org/10.1145/1454008.1454027}>. PFP distributes computation in such a way that each worker executes an
```
This line seems exceeding the length limit.

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+#' @param numPartitions Number of partitions used for fitting.
+#' @param ... additional argument(s) passed to the method.
+#' @return \code{spark.fpGrowth} returns a fitted FPGrowth model.
+#' 
```
Other APIs do not have blank line here. I think we should be consistent.

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+          })
+
+# Get frequent itemsets.
+#' @param object a fitted FPGrowth model.
```
add blank line. See other examples.

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+# Get frequent itemsets.
+#' @param object a fitted FPGrowth model.
+#' @return A DataFrame with frequent itemsets.
+#' 
```
no blank line here.

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+          })
+
+# Get association rules.
+#' @return A DataFrame with association rules.
```
Add blank line.

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+          })
+
+#  Makes predictions based on generated association rules
+#' @param newData a SparkDataFrame for testing.
```
Add blank line

wangmiao1981 commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,152 @@
+          })
+
+#  Saves the FPGrowth model to the output path.
+#' @param path the directory where the model is saved.
```
add blank line

wangmiao1981 commented on mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala:
```diff
@@ -0,0 +1,87 @@
+private[r] object FPGrowthWrapper extends MLReadable[FPGrowthWrapper] {
+
+  def fit(
+         data: DataFrame,
```
alignment

wangmiao1981 commented on mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala:
```diff
@@ -0,0 +1,87 @@
+    }
+  }
+
+    class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter {
```
indentation seems incorrect here and above line.

SparkQA commented:
**[Test build #74744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74744/testReport)** for PR 17170 at commit [`12ad3ff`](https://github.com/apache/spark/commit/12ad3ff10bfebf5dabef210b4678e045bd5c673a).

SparkQA commented:
**[Test build #74745 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74745/testReport)** for PR 17170 at commit [`521ec38`](https://github.com/apache/spark/commit/521ec387d7d02f298dacdee0629d87a8800f9f6f).

SparkQA commented:
**[Test build #74744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74744/testReport)** for PR 17170 at commit [`12ad3ff`](https://github.com/apache/spark/commit/12ad3ff10bfebf5dabef210b4678e045bd5c673a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74744/
Test FAILed.

SparkQA commented:
**[Test build #74745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74745/testReport)** for PR 17170 at commit [`521ec38`](https://github.com/apache/spark/commit/521ec387d7d02f298dacdee0629d87a8800f9f6f).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74745/
Test FAILed.

SparkQA commented:
**[Test build #74746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74746/testReport)** for PR 17170 at commit [`89a5815`](https://github.com/apache/spark/commit/89a5815471069298d2fbbc12ca5b4d3cbf8c98c9).

SparkQA commented:
**[Test build #74746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74746/testReport)** for PR 17170 at commit [`89a5815`](https://github.com/apache/spark/commit/89a5815471069298d2fbbc12ca5b4d3cbf8c98c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74746/
Test PASSed.

SparkQA commented:
**[Test build #74747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74747/testReport)** for PR 17170 at commit [`7635afc`](https://github.com/apache/spark/commit/7635afc107cc1778a2fbf6861a5ebc164eab421e).

SparkQA commented:
**[Test build #74747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74747/testReport)** for PR 17170 at commit [`7635afc`](https://github.com/apache/spark/commit/7635afc107cc1778a2fbf6861a5ebc164eab421e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74747/
Test PASSed.

felixcheung commented:
I suppose we should wait on #17321 before acting on this PR?

zero323 commented:
@felixcheung It is completely up to you. I'll have to patch one or another (and also #17218) and at the end of the day it doesn't make much difference.

felixcheung commented:
Could you update this PR to have the parameter  itemsCol
And remove predictionCol (if I recall, we don't expose that in the R API for other models either)

SparkQA commented:
**[Test build #74823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74823/testReport)** for PR 17170 at commit [`dd7822e`](https://github.com/apache/spark/commit/dd7822e36ce1b573f0781182ffd713a50f0563e1).

SparkQA commented:
**[Test build #74823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74823/testReport)** for PR 17170 at commit [`dd7822e`](https://github.com/apache/spark/commit/dd7822e36ce1b573f0781182ffd713a50f0563e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74823/
Test PASSed.

zero323 commented:
Done. I marked it as blocked by SPARK-19899 and leave it to be retested once #17321 is merged.

SparkQA commented:
**[Test build #74826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74826/testReport)** for PR 17170 at commit [`706514d`](https://github.com/apache/spark/commit/706514da26107ef25bef028e2143fa0a09e5cc19).

SparkQA commented:
**[Test build #74826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74826/testReport)** for PR 17170 at commit [`706514d`](https://github.com/apache/spark/commit/706514da26107ef25bef028e2143fa0a09e5cc19).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74826/
Test FAILed.

zero323 commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #74896 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74896/testReport)** for PR 17170 at commit [`706514d`](https://github.com/apache/spark/commit/706514da26107ef25bef028e2143fa0a09e5cc19).

SparkQA commented:
**[Test build #74896 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74896/testReport)** for PR 17170 at commit [`706514d`](https://github.com/apache/spark/commit/706514da26107ef25bef028e2143fa0a09e5cc19).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74896/
Test PASSed.

felixcheung commented:
this is the error from appveyor build
```
[00:16:14] [ERROR] C:\projects\spark\mllib\src\main\scala\org\apache\spark\ml\r\FPGrowthWrapper.scala:50: value setItemsCol is not a member of org.apache.spark.ml.fpm.FPGrowth
[00:16:14] possible cause: maybe a semicolon is missing before `value setItemsCol'?
[00:16:14] [ERROR]       .setItemsCol(itemsCol)
[00:16:14] [ERROR]        ^
[00:16:29] [ERROR] one error found
```

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,153 @@
+#' @note FPGrowthModel since 2.2.0
+setClass("FPGrowthModel", slots = list(jobj = "jobj"))
+
+#' FPGrowth
```
I think we discussed this - let's make it `FP-Growth` or `Frequent Pattern Mining` (https://spark.apache.org/docs/latest/mllib-frequent-pattern-mining.html) as the title

felixcheung replied:
was https://github.com/apache/spark/pull/17170#discussion_r104736398

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,153 @@
+#' 
+#' A parallel FP-growth algorithm to mine frequent itemsets. The algorithm is described in
+#' Li et al., PFP: Parallel FP-Growth for Query
+#' Recommendation <\url{http://dx.doi.org/10.1145/1454008.1454027}>. 
```
can you check if this generate the doc properly
`<\url{http://dx.doi.org/10.1145/1454008.1454027}>`
generally it should be 
`\href{http://...}{Text}`

zero323 replied:
It does render the link as expected, but linking ML docs is indeed a better choice.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,153 @@
+#' PFP distributes computation in such a way that each worker executes an
+#' independent group of mining tasks. The FP-Growth algorithm is described in
+#' Han et al., Mining frequent patterns without
+#' candidate generation <\url{http://dx.doi.org/10.1145/335191.335372}>.
```
ditto here for url.
In fact, I'm not sure we need to include all the links here but instead link to 
https://spark.apache.org/docs/latest/mllib-frequent-pattern-mining.html

zero323 replied:
Sounds good. I'll link the docs.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,153 @@
+#' another_model <- spark.fpGrowth(data, minSupport = 0.1, minConfidence = 0.5
+#'                                 itemsCol = "baskets", numPartitions = 10)
+#' }
+#' @references \url{http://en.wikipedia.org/wiki/Association_rule_learning}
```
we don't generally use this tag. Do you want to move to @seealso, or just link to in the description above

zero323 replied:
I'll remove it completely and just link to the docs.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,153 @@
+#' @note spark.fpGrowth since 2.2.0
+setMethod("spark.fpGrowth", signature(data = "SparkDataFrame"),
+          function(data, minSupport = 0.3, minConfidence = 0.8,
+                   itemsCol = "items", numPartitions = -1) {
```
`numPartitions` by default is not set in Scala - let's default this to NULL instead here
(but do not as.integer if value is NULL - something like
numPartitions <- if (is.null(numPartitions)) NULL else as.integer(numPartitions)

zero323 replied:
Correct if I am wrong but this cannot be done like this. If we want to default to `NULL` (I am not fond of this idea) we have to pass argument as a `character` / `String` and parse it once in JVM.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala:
```diff
@@ -0,0 +1,86 @@
+      .setMinConfidence(minConfidence)
+      .setItemsCol(itemsCol)
+
+    if (numPartitions != null && numPartitions > 0) {
```
given the earlier suggestion, we should also check numPartition > 0 in R before passing to here

felixcheung replied:
and this comment https://github.com/apache/spark/pull/17170#discussion_r107011745

felixcheung replied:
and https://github.com/apache/spark/pull/17170#discussion_r107349375

zero323 replied:
If you feel it is necessary. Personally I wanted to treat any non-strictly positive number as `null`.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala:
```diff
@@ -0,0 +1,86 @@
+  class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter {
+    override protected def saveImpl(path: String): Unit = {
+      val modelPath = new Path(path, "model").toString
+      val rMetadataPath = new Path(path, "rMetadata").toString
```
anything else we could add as metadata that is not in the model already?

zero323 replied:
I don't think so. Model captures all the parameters.

felixcheung reviewed (changes requested):
generally good, a few last comments

SparkQA commented:
**[Test build #75007 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75007/testReport)** for PR 17170 at commit [`999aa7a`](https://github.com/apache/spark/commit/999aa7a9c483d4a89f8457092850831be3a05093).

SparkQA commented:
**[Test build #75007 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75007/testReport)** for PR 17170 at commit [`999aa7a`](https://github.com/apache/spark/commit/999aa7a9c483d4a89f8457092850831be3a05093).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `  class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75007/
Test FAILed.

SparkQA commented:
**[Test build #75008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75008/testReport)** for PR 17170 at commit [`6522916`](https://github.com/apache/spark/commit/65229163721475f7769387d3e4ba912e570cecc3).

SparkQA commented:
**[Test build #75008 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75008/testReport)** for PR 17170 at commit [`6522916`](https://github.com/apache/spark/commit/65229163721475f7769387d3e4ba912e570cecc3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FPGrowthWrapperReader extends MLReader[FPGrowthWrapper] `
  * `  class FPGrowthWrapperWriter(instance: FPGrowthWrapper) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75008/
Test PASSed.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,148 @@
+              stop("minConfidence should be a number [0, 1].")
+            }
+
+            numPartitions <- if (is.null(numPartitions)) NULL else as.integer(numPartitions)
```
as this https://github.com/apache/spark/pull/17170/commits/65229163721475f7769387d3e4ba912e570cecc3#r107011745 we should check numPartitions too?
How about changing it to
```
if (!is.null(numPartitions)) {
  numPartitions <- as.integer(numPartitions)
  stopifnot(numPartitions > 0)
}
```

felixcheung commented:
also would you be updating the R vignettes, ML programming guide and example?

zero323 commented:
Let's make it a separate task. For ML guide we have to wait for #17130 anyway.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -0,0 +1,148 @@
+
+# Get association rules.
+
+#' @return A DataFrame with association rules.
```
let's document the list of column like in Python: https://github.com/apache/spark/pull/17218/files#diff-b6dbf16870bd2cca9b4140df8aebd681R121

for reference, see https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L249

felixcheung commented on R/pkg/inst/tests/testthat/test_mllib_fpm.R:
```diff
@@ -0,0 +1,76 @@
+    "1,3"
+  ))), "split(items, ',') as items")
+
+  model <- spark.fpGrowth(data, minSupport = 0.3, minConfidence = 0.8, numPartitions = 1)
```
we need to add a test when numPartitions is not set...

zero323 committed:
Stop if numPartitions is not null and not strictly positive

zero323 committed:
Add test for training without numPartitions

SparkQA commented:
**[Test build #75275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75275/testReport)** for PR 17170 at commit [`2f49f98`](https://github.com/apache/spark/commit/2f49f9888a56221f609e07574af5e78753211359).

SparkQA commented:
**[Test build #75275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75275/testReport)** for PR 17170 at commit [`2f49f98`](https://github.com/apache/spark/commit/2f49f9888a56221f609e07574af5e78753211359).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75275/
Test FAILed.

SparkQA commented:
**[Test build #75276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75276/testReport)** for PR 17170 at commit [`8f0e578`](https://github.com/apache/spark/commit/8f0e5787abffe367da7ae96d3c2f6b517b89ffb4).

SparkQA commented:
**[Test build #75276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75276/testReport)** for PR 17170 at commit [`8f0e578`](https://github.com/apache/spark/commit/8f0e5787abffe367da7ae96d3c2f6b517b89ffb4).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75276/
Test FAILed.

zero323 commented:
Jenkins retest this please.

zero323 commented:
@felixcheung Looks like some issue with the structured streaming: https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75276/

felixcheung commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #75293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75293/testReport)** for PR 17170 at commit [`8f0e578`](https://github.com/apache/spark/commit/8f0e5787abffe367da7ae96d3c2f6b517b89ffb4).

felixcheung commented:
@zero323 how about https://github.com/apache/spark/pull/17170#discussion_r108079970

SparkQA commented:
**[Test build #75293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75293/testReport)** for PR 17170 at commit [`8f0e578`](https://github.com/apache/spark/commit/8f0e5787abffe367da7ae96d3c2f6b517b89ffb4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75293/
Test PASSed.

zero323 committed:
Add column description for freqItemsets and associationRules

SparkQA commented:
**[Test build #75320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75320/testReport)** for PR 17170 at commit [`797d68d`](https://github.com/apache/spark/commit/797d68d517dbeacb61e0577e243066f5a8812c15).

SparkQA commented:
**[Test build #75320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75320/testReport)** for PR 17170 at commit [`797d68d`](https://github.com/apache/spark/commit/797d68d517dbeacb61e0577e243066f5a8812c15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75320/
Test PASSed.

felixcheung commented on R/pkg/R/mllib_fpm.R:
```diff
@@ -99,7 +99,10 @@ setMethod("spark.fpGrowth", signature(data = "SparkDataFrame"),
 
 #' @param object a fitted FPGrowth model.
-#' @return A DataFrame with frequent itemsets.
+#' @return A \code{DataFrame} with frequent itemsets.
```
Actually, sorry we need to change `DataFrame` to `SparkDataFrame` in R

zero323 committed:
Correct names DataFrame -> SparkDataFrame

SparkQA commented:
**[Test build #75366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75366/testReport)** for PR 17170 at commit [`64c07aa`](https://github.com/apache/spark/commit/64c07aaa0d538c6d0fe01a8fe831e11194603e22).

SparkQA commented:
**[Test build #75366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75366/testReport)** for PR 17170 at commit [`64c07aa`](https://github.com/apache/spark/commit/64c07aaa0d538c6d0fe01a8fe831e11194603e22).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75366/
Test PASSed.

felixcheung commented:
merged to master.
@zero323 could you follow up with vignettes and programming guide update please - we need them for the 2.2.0 release.

zero323 commented:
Of course. Do we have / need a JIRA ticket for that?

felixcheung commented:
for this, it's optional, but I opened one for tracking purpose https://issues.apache.org/jira/browse/SPARK-20208?filter=12333531
